src/artm/regularizer/topic_segmentation_ptdw.cc

```diff
@@ -16,107 +17,272 @@ namespace regularizer {
+      norm_sum += weight * weight;
+      for (int t = 0; t < num_topics; t++) {
+        if (is_background_topic[t])
+          continue;
+        sen_subj[t] += copy_ptdw(i, t) * weight;
+      }
+    }
+    for (int t = 0; t < num_topics; t++) {
+      if (is_background_topic[t])
+        continue;
+      if (norm_sum != 0)
+        sen_subj[t] /= norm_sum;
     }
-    left_weights += 1 - background_probability[i - 1];
-    right_weights -= 1 - background_probability[i - 1];
-    if (i <= local_token_size - h) {
-      for (int k = 0; k < topic_size; ++k) {
-        right_distribution[k] += copy_ptdw(i + h - 1, k) * (1 - background_probability[i + h - 1]);
+    for (int i = sen_begin; i < sen_end; i++) {
+      float TMP = 0.0f;
+      float weight = 1.0f - background_probability[i];
+      //weight = 1.0f;
+      for (int t = 0; t < num_topics; t++) {
+        if (is_background_topic[t])
+          continue;
+        if (copy_ptdw(i, t) != 0 && sen_subj[t] != 0)
+          TMP += copy_ptdw(i, t) / sen_subj[t];
+      }
+      float sum = 0.0f;
+      float non_backs = 1.0f;
+
+      for (int t = 0; t < num_topics; t++) {
+        if (copy_ptdw(i, t) != 0 && norm_sum != 0) {
+          if (is_background_topic[t]) {
+            //(*ptdw)(i, t) = copy_ptdw(i, t) * (1.0f + tau * weight * TMP);
+          } else {
+           (*ptdw)(i, t) = copy_ptdw(i, t) * (1.0f - tau * (weight / norm_sum) * (1.0f / sen_subj[t] - TMP));
+          }
+        }
+        if ((*ptdw)(i, t) < 0.0f)
+          (*ptdw)(i, t) = 0.0f;
+        if (! is_background_topic[t])
+          sum += (*ptdw)(i, t);
+        else
+          non_backs -= (*ptdw)(i, t);
+
+      }
+      for (int t = 0; t < num_topics; t++) {
+        if ((*ptdw)(i, t) != 0 && !is_background_topic[t])
+          (*ptdw)(i, t) = non_backs * (*ptdw)(i, t) / sum;
+
       }
-      right_weights += 1 - background_probability[i + h - 1];
     }
-    if (i > h) {
-      for (int k = 0; k < topic_size; ++k) {
-        left_distribution[k] -= copy_ptdw(i - h - 1, k) * (1 - background_probability[i - h - 1]);
+    if (!handling_first_sen) {
+      float norm_prev = 0.0f;
+      float norm_cur = 0.0f;
+      float dot = 0.0f;
+      for (int t = 0; t < num_topics; t++) {
+        float sen_subj_t = sen_subj[t];
+        float prev_sen_subj_t = prev_sen_subject[t];
+        // if (!handling_first_sen && sen_len < average_len / 2.0)
+        //   sen_subj_t = (1.0f - alpha) * sen_subj[t] + alpha * prev_sen_subject[t];
+        // else if (!handling_first_sen && prev_seg_len < average_len / 2.0)
+        //   prev_sen_subj_t = (1.0f - alpha) * prev_sen_subject[t] + alpha * sen_subj[t];
+
+        dot += prev_sen_subj_t * sen_subj_t;
+        norm_cur += sen_subj_t * sen_subj_t;
+        norm_prev += prev_sen_subj_t * prev_sen_subj_t;
+      }
+      norm_prev = std::sqrt(norm_prev);
+      norm_cur = std::sqrt(norm_cur);
+      float dist = 1.0f;
+      if (norm_prev > 1e-5 && norm_cur > 1e-5) {
```

Comment:
move all eps const into const variables in config or at least in the head of this file

Reply:
and 1e-5f